File: internal/browserprotocol/transcript.go
// Package browserprotocol contains the small, framework-neutral cryptographic
// pieces shared by the browser transport and its clients.
package browserprotocol

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"math/big"
	"unicode/utf8"
)

const (
	ProtocolVersion = uint16(2)
	DaemonIDSize    = 16
	BootIDSize      = 16
	NonceSize       = 32
	ChallengeSize   = 32
	ClientIDSize    = 16
	PublicKeySize   = 65
	SignatureSize   = 64
	MaxTextBytes    = 4096
)

var (
	ErrInvalidLength    = errors.New("browser protocol: invalid field length")
	ErrInvalidText      = errors.New("browser protocol: invalid host or origin")
	ErrInvalidPublicKey = errors.New("browser protocol: invalid P-256 public key")
	ErrInvalidSignature = errors.New("browser protocol: invalid ECDSA signature")
)

// PairTranscript contains the values bound by a pairing proof. Byte slices
// are used deliberately: callers crossing the wire must be checked for the
// exact fixed lengths instead of being able to construct a malformed array.
type PairTranscript struct {
	DaemonID        []byte
	BootID          []byte
	ConnectionNonce []byte
	Challenge       []byte
	PublicKeySEC1   []byte
	ValidatedHost   string
	ValidatedOrigin string
}

// AuthTranscript contains the values bound by an existing-client proof.
type AuthTranscript struct {
	DaemonID        []byte
	BootID          []byte
	ConnectionNonce []byte
	ClientID        []byte
	ValidatedHost   string
	ValidatedOrigin string
}

const (
	pairDomain = "dark-factory/browser/v2/pair\x00"
	authDomain = "dark-factory/browser/v2/auth\x00"
)

// BuildPairTranscript returns the exact bytes signed by a browser while
// pairing. The domain is intentionally unlength-prefixed; every field after
// the version has a u32 big-endian byte length.
func BuildPairTranscript(input PairTranscript) ([]byte, error) {
	return buildTranscript(pairDomain, [][]byte{
		input.DaemonID,
		input.BootID,
		input.ConnectionNonce,
		input.Challenge,
		input.PublicKeySEC1,
		[]byte(input.ValidatedHost),
		[]byte(input.ValidatedOrigin),
	}, []int{DaemonIDSize, BootIDSize, NonceSize, ChallengeSize, PublicKeySize, 0, 0})
}

// BuildAuthTranscript returns the exact bytes signed by a browser while
// authenticating an existing client.
func BuildAuthTranscript(input AuthTranscript) ([]byte, error) {
	return buildTranscript(authDomain, [][]byte{
		input.DaemonID,
		input.BootID,
		input.ConnectionNonce,
		input.ClientID,
		[]byte(input.ValidatedHost),
		[]byte(input.ValidatedOrigin),
	}, []int{DaemonIDSize, BootIDSize, NonceSize, ClientIDSize, 0, 0})
}

func buildTranscript(domain string, fields [][]byte, fixed []int) ([]byte, error) {
	if len(fields) != len(fixed) {
		return nil, fmt.Errorf("%w: field count", ErrInvalidLength)
	}
	result := make([]byte, 0, len(domain)+2+len(fields)*4)
	result = append(result, domain...)
	var version [2]byte
	binary.BigEndian.PutUint16(version[:], ProtocolVersion)
	result = append(result, version[:]...)
	for i, field := range fields {
		if fixed[i] != 0 && len(field) != fixed[i] {
			return nil, fmt.Errorf("%w: field %d is %d bytes, want %d", ErrInvalidLength, i, len(field), fixed[i])
		}
		if fixed[i] == 0 {
			if err := validateText(field); err != nil {
				return nil, fmt.Errorf("field %d: %w", i, err)
			}
		}
		if uint64(len(field)) > uint64(^uint32(0)) {
			return nil, fmt.Errorf("%w: field %d exceeds u32", ErrInvalidLength, i)
		}
		var length [4]byte
		binary.BigEndian.PutUint32(length[:], uint32(len(field)))
		result = append(result, length[:]...)
		result = append(result, field...)
	}
	return result, nil
}

func validateText(value []byte) error {
	if len(value) == 0 || len(value) > MaxTextBytes || !utf8.Valid(value) {
		return ErrInvalidText
	}
	for _, b := range value {
		if b == 0 {
			return ErrInvalidText
		}
	}
	return nil
}

// VerifySignature verifies an IEEE-P1363 P-256 signature over transcript.
// The complete transcript is hashed exactly once here. ASN.1 and alternate
// public-key encodings are intentionally not accepted.
func VerifySignature(publicKeySEC1, signature, transcript []byte) error {
	if len(signature) != SignatureSize {
		return fmt.Errorf("%w: signature is %d bytes, want %d", ErrInvalidSignature, len(signature), SignatureSize)
	}
	if len(publicKeySEC1) != PublicKeySize || publicKeySEC1[0] != 4 {
		return fmt.Errorf("%w: require 65-byte uncompressed SEC1 point", ErrInvalidPublicKey)
	}
	x, y := elliptic.Unmarshal(elliptic.P256(), publicKeySEC1)
	if x == nil || y == nil || !elliptic.P256().IsOnCurve(x, y) {
		return ErrInvalidPublicKey
	}
	curve := elliptic.P256()
	order := curve.Params().N
	r := new(big.Int).SetBytes(signature[:32])
	s := new(big.Int).SetBytes(signature[32:])
	if r.Sign() <= 0 || r.Cmp(order) >= 0 || s.Sign() <= 0 || s.Cmp(order) >= 0 {
		return ErrInvalidSignature
	}
	digest := sha256.Sum256(transcript)
	if !ecdsa.Verify(&ecdsa.PublicKey{Curve: curve, X: x, Y: y}, digest[:], r, s) {
		return ErrInvalidSignature
	}
	return nil
}

File: web/packages/client/src/transcript.ts
import { malformed, normalizeBoundary, ProtocolError } from "./errors.js";
import { PROTOCOL_VERSION } from "./manifest.js";

const PROTOCOL = new TextEncoder().encode("dark-factory/browser/v2/");
const PAIR = new TextEncoder().encode("pair\0");
const AUTH = new TextEncoder().encode("auth\0");

export type PairTranscriptInput = {
  daemon_id: string; boot_id: string; connection_nonce: string; challenge: string;
  public_key_sec1: string; host: string; origin: string;
};
export type AuthTranscriptInput = {
  daemon_id: string; boot_id: string; connection_nonce: string; client_id: string;
  host: string; origin: string;
};

export function hexBytes(value: string, bytes?: number): Uint8Array {
  return normalizeBoundary(() => {
    if (typeof value !== "string" || (bytes !== undefined && (!Number.isSafeInteger(bytes) || bytes < 0)) || value.length % 2 !== 0 || value !== value.toLowerCase() || !/^[0-9a-f]*$/.test(value)) malformed();
    if (bytes !== undefined && value.length !== bytes * 2) malformed();
    const result = new Uint8Array(value.length / 2);
    for (let i = 0; i < result.length; i++) result[i] = Number.parseInt(value.slice(i * 2, i * 2 + 2), 16);
    return result;
  });
}
function text(value: string): Uint8Array {
  for (const character of value) {
    const code = character.codePointAt(0) ?? 0;
    if (code >= 0xd800 && code <= 0xdfff) malformed();
  }
  const result = new TextEncoder().encode(value);
  if (result.length === 0 || result.length > 4096 || result.includes(0)) malformed();
  return result;
}
function fixed(value: string, bytes: number): Uint8Array { return hexBytes(value, bytes); }
function publicKey(value: string): Uint8Array { const result = fixed(value, 65); if (result[0] !== 4) malformed(); return result; }
function transcript(domain: Uint8Array, fields: Array<{ value: Uint8Array; fixed?: number }>): Uint8Array {
  const parts = [PROTOCOL, domain, new Uint8Array([PROTOCOL_VERSION >>> 8, PROTOCOL_VERSION & 0xff])];
  let size = parts.reduce((n, p) => n + p.length, 0);
  for (const field of fields) { if (field.fixed !== undefined && field.value.length !== field.fixed) malformed(); const length = new Uint8Array(4); new DataView(length.buffer).setUint32(0, field.value.length); parts.push(length, field.value); size += 4 + field.value.length; }
  const result = new Uint8Array(size); let offset = 0; for (const part of parts) { result.set(part, offset); offset += part.length; } return result;
}
export function buildPairTranscript(input: PairTranscriptInput): Uint8Array {
  return normalizeBoundary(() => {
    if (typeof input !== "object" || input === null) malformed();
    return transcript(PAIR, [
      { value: fixed(input.daemon_id, 16), fixed: 16 }, { value: fixed(input.boot_id, 16), fixed: 16 }, { value: fixed(input.connection_nonce, 32), fixed: 32 },
      { value: fixed(input.challenge, 32), fixed: 32 }, { value: publicKey(input.public_key_sec1), fixed: 65 }, { value: text(input.host) }, { value: text(input.origin) },
    ]);
  });
}
export function buildAuthTranscript(input: AuthTranscriptInput): Uint8Array {
  return normalizeBoundary(() => {
    if (typeof input !== "object" || input === null) malformed();
    return transcript(AUTH, [
      { value: fixed(input.daemon_id, 16), fixed: 16 }, { value: fixed(input.boot_id, 16), fixed: 16 }, { value: fixed(input.connection_nonce, 32), fixed: 32 },
      { value: fixed(input.client_id, 16), fixed: 16 }, { value: text(input.host) }, { value: text(input.origin) },
    ]);
  });
}

/** Verify the raw IEEE-P1363 P-256 signature used by browser v1. */
export async function verifyP256Signature(publicKeySEC1: Uint8Array, signature: Uint8Array, signed: Uint8Array): Promise<boolean> {
  if (!(publicKeySEC1 instanceof Uint8Array) || !(signature instanceof Uint8Array) || !(signed instanceof Uint8Array)) throw new ProtocolError("malformed");
  if (publicKeySEC1.length !== 65 || publicKeySEC1[0] !== 4 || signature.length !== 64) return false;
  try {
    const key = await globalThis.crypto.subtle.importKey("raw", publicKeySEC1 as unknown as BufferSource, { name: "ECDSA", namedCurve: "P-256" }, false, ["verify"]);
    return await globalThis.crypto.subtle.verify({ name: "ECDSA", hash: "SHA-256" }, key, signature as unknown as BufferSource, signed as unknown as BufferSource);
  } catch { return false; }
}
